flexmeasures/data/models/planning/storage.py

```diff
@@ -198,14 +199,40 @@ def _prepare(self, skip_validation: bool = False) -> tuple:  # noqa: C901
             soc_min,
         )
 
+        consumption_capacity = self.flex_model.get("consumption_capacity")
+        production_capacity = self.flex_model.get("production_capacity")
+
         if sensor.get_attribute("is_strictly_non_positive"):
             device_constraints[0]["derivative min"] = 0
         else:
-            device_constraints[0]["derivative min"] = power_capacity_in_mw * -1
+            device_constraints[0]["derivative min"] = (
+                -1
+            ) * get_series_from_sensor_or_quantity(
+                production_capacity,
+                sensor,
+                "production_capacity",
+                query_window=(start, end),
+                resolution=resolution,
+                beliefs_before=belief_time,
+            ).fillna(
+                power_capacity_in_mw
+            )
+
         if sensor.get_attribute("is_strictly_non_negative"):
             device_constraints[0]["derivative max"] = 0
         else:
-            device_constraints[0]["derivative max"] = power_capacity_in_mw
+            device_constraints[0][
+                "derivative max"
+            ] = get_series_from_sensor_or_quantity(
+                consumption_capacity,
+                sensor,
+                "consumption_capacity",
+                query_window=(start, end),
+                resolution=resolution,
+                beliefs_before=belief_time,
+            ).fillna(
+                power_capacity_in_mw
+            )
```

Comment:
This is only supposed to happen when all values are NaN, right?

Reply:
I think that the power capacity of the device should default to a known value. Actually, here we should probably clip the time series value to power_capacity_in_mw.

Reply:
This is where we would do something like a `nanmin()`, right? As we did elsewhere.

Reply:
I think first stacking horizontally the one-way capacity with the two-way capacity and then apply nanmin on row-wise.

Reply:
Exactly, or `.clip(power_capacity).fillna(power_capacity)`, whichever is fastest. So then maybe the util function becomes cleaner by not needing the `default_value` anymore?

Also consider splitting up the util function into `get_series_from_sensor_or_quantity`, which only does what its name suggests (i.e. doesn't deal with default fallback values), and a new util function, that deals with filling in the blanks.

Reply:
Good idea!